src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -42,17 +44,39 @@ internal void WriteAsV7(Stream archiveStream, Span<byte> buffer)
             }
         }
 
+        // Asynchronously writes the current header as a V7 entry into the archive stream and returns the value of the final checksum.
+        internal async Task<int> WriteAsV7Async(Stream archiveStream, Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            long actualLength = GetTotalDataBytesToWrite();
+            TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarEntryFormat.V7);
+
+            int checksum = WriteNameMemory(buffer, out _);
+            checksum += WriteCommonFields(buffer.Span, actualLength, actualEntryType);
+            int finalChecksum = WriteChecksum(checksum, buffer.Span);
+
+            await archiveStream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);
+
+            if (_dataStream != null)
+            {
+                await WriteDataAsync(archiveStream, _dataStream, actualLength, cancellationToken).ConfigureAwait(false);
+            }
+
+            return finalChecksum;
```

Comment:
It seems that we could reduce code duplication here by moving part of the common logic to a separate method:

```cs
private long WriteCommonFieldsAndChecksumAsV7(Span<byte> buffer)
{
    long actualLength = GetTotalDataBytesToWrite();
    TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarEntryFormat.V7);

    int checksum = WriteName(buffer, out _);
    int checksum = WriteNameSpan(buffer, out _);
    checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
    WriteChecksum(checksum, buffer);
    _checksum = WriteChecksum(checksum, buffer);
    
    return actualLength;
}

internal void WriteAsV7(Stream archiveStream, Span<byte> buffer)
{
    long actualLength = WriteCommonFieldsAndChecksumAsV7(buffer);

    archiveStream.Write(buffer);

    if (_dataStream != null)
    {
        WriteData(archiveStream, _dataStream, actualLength);
    }
}

internal async Task<int> WriteAsV7Async(Stream archiveStream, Memory<byte> buffer, CancellationToken cancellationToken)
{
    cancellationToken.ThrowIfCancellationRequested();

    long actualLength = WriteCommonFieldsAndChecksumAsV7(buffer.Span);

    await archiveStream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);

    if (_dataStream != null)
    {
        await WriteDataAsync(archiveStream, _dataStream, actualLength, cancellationToken).ConfigureAwait(false);
    }

    return _checksum;
}
```